2022/Q3/artifacts/PoS-tla/MC_Staking.tla

```diff
@@ -160,34 +171,136 @@ HighCoverage(trace) ==
+ *      that voting power.
+ * (ii) If evidence is found in time and there are no repeated infractions within 
+ *      an epoch window, then the proof-of-stake model guarantees slashing.
+ *
+ * Invariant #6 verifies property (i) for the case in which there is no slashing.
+ * This is the nly case in which we have totalDeltas = bonds.
+ *
+ * Invariant #2 verifies property (ii).
+*)
+
+(*
+ * Invariant #1
+ * The users' balance is always greater or equal to zero
+*)
+
+BalanceAlwaysPositive == 
+    \A user \in UserAddrs: balanceOf[user] >= 0
 
-\* Invariant #1
-\* the validator's voting power at a given epoch is equal to the total amount of
-\* tokens delegated to that validator
 
-(* ValDelegatedFold(set, val) == LET SumDelegated(p,q) == p + delegated[1, q, val]
-                              IN ApaFoldSet( SumDelegated, 0, set )
+(* 
+ * Invariant #2
+ * posAccount is always greater or equal to zero
+ *
+ * This implies the following:
+ * Assume that a validator misbehaves at epoch e and that evidence is found at an epoch e’ <= e + unbonding_length.
+ * Assume also that the validator only misbehaves once. Then, any user whose bonds (including the validators self-bonds)
+ * were contributing to the misbehaving validator's voting power are slashed when withdrawing any of those bonds.
+ *
+ * The logic is the following:
+ * When we slash a validator (reduce its voting power) at the end of an epoch, we move tokens from the posAccount to
+ * the slashPool. Thus, in the posAccount, we only have the bonded tokens that are withdrawable: this means, the bonded
+ * tokens minus the slashsable amount. If a user is able to withdraw tokens that were contributing to the voting 
+ * power of the misbehaving validator, then in a execution were all toekns are withdrawn, the posAccount should go
+ * negative.
+*)
+
+PoSAccountAlwaysPositive == 
+    posAccount >= 0
+
+\* Auxiliary functions for Invariant #3.
+
+bondsWithOpenEnd == LET
+                    \* @type: (Set(BOND), <<ADDR, ADDR>>) => Set(BOND);   
+                    F(set, pair) == set \union { bond \in bonded[pair[1], pair[2]]: bond.end = -1 }
+                    IN ApaFoldSet(F, {}, UserAddrs \X ValidatorAddrs)
+
+foldUnbonds == LET
+               \* @type: (Set(UNBOND), <<ADDR, ADDR>>) => Set(UNBOND);   
+               F(set, pair) == set \union unbonded[pair[1], pair[2]]
+               IN ApaFoldSet(F, {}, UserAddrs \X ValidatorAddrs)
+
+(* 
+ * Invariant #3
+ * First bullet from https://github.com/informalsystems/partnership-heliax/issues/25
+ *
+ * The amount that is self-bonded and delegated tokens is exactly equal to the amount that can be withdrawn back from PoS.
+ * In other words, if all the bonds are withdrawn, the PoS balance must be equal to zero.
+*)
+
+posAccountZero ==
+    bondsWithOpenEnd = {} /\ foldUnbonds = {} => posAccount = 0
+
+\* Auxiliary functions for Invariants #4, #5 and #6.            
+
+SumBondsUser(user, val) == LET
```

Comment:
Since `SumBondsUser` is only used as part of `TotalSumBonds`, I understand you splitted them for readability.
However, aren't we splitting a single loop into many (affecting efficiency)?

Reply:
`SumBondsUser` iterates over users and `TotalSumBonds` over the bonds of a given validator and user, so I think it is not a single loop.